### PR TITLE
558 workspaces

### LINKED ISSUE #558: Use Rust workspaces
# Motivation:
- useful for plugin development, e.g. shared C types, plugin sdk, plugin examples
- might improve build time

# To get plugins going:
- setup workspaces
- move all current code into a remote_server crate

# Next steps:
- crate for shared domain types (e.g. shared between repo and service layer)
- crate for repos
- crate for services
- crate for gql

### PR DESCRIPTION
**Pull request by clemens-msupply**
The workspace has the following crates:
- domain
- repository
- service
- graphql
- server

I tried to minimize the changes unrelated to the workspace migration and there are a couple of things that could be refactored in the future, (e.g. another utils crate?) However, some changes where necessary to keep the server working:

- Update readme, e.g. say that you should be in the server directory for the migration commands and to run the server
- update graphql_export so that it can run from the project root and from the graphql crate (it tries to auto detect from where it is running and where to generate the output, thought this could avoids some confusion...)
- Removed RepositoryRegistry and only store the StorageConnectionManager in the app data (this simplified the workspace transition and RepositoryRegistry was only holding StorageConnectionManager anyway...)
- Inline a handful of `from`s (3 or 4 if I remember correctly) They caused errors like below and it was easier to inline them than to workaround the tuple problem:
```
error[E0117]: only traits defined in the current crate can be implemented for arbitrary types
  --> server/src/database/repository/./diesel/item_query.rs:25:1
   |
25 | impl From<ItemAndMasterList> for Item {
   | ^^^^^-----------------------^^^^^----
   | |    |                           |
   | |    |                           `domain::item::Item` is not defined in the current crate
   | |    this is not defined in the current crate because tuples are always foreign
   | impl doesn't use only types from inside the current crate
   |
   = note: define and implement a trait or new type instead
```

EDIT: I also removed the "mock" feature, should I put it back in? Was late last night and I spontaneously decided we are not using it and it can be removed. 

Closes #558

### CODE REVIEW

File: repository/src/repository/diesel/item_query.rs [hunk 1]
```diff
@@ -85,7 +71,18 @@ impl<'a> ItemQueryRepository<'a> {
             .limit(pagination.limit as i64)
             .load::<ItemAndMasterList>(&self.connection.connection)?;
 
-        Ok(result.into_iter().map(Item::from).collect())
+        Ok(result
+            .into_iter()
+            .map(
```
Comment (andreievg) [inline repository/src/repository/diesel/item_query.rs:76]: I really think this should be a method.

fn toDomain( .. )

.map(toDomain)

Comment (clemens-msupply) [reply repository/src/repository/diesel/item_query.rs:76]: Actually like it because the whole method is not to big and having it unlined leaves all code in one place.

Will give the to_domain method a try to see how it looks. Or would comment help? e.g.

``` rust 
       // transform DB results to domain objects
       Ok(result
            .into_iter()
            .map(
```

File: README.md [hunk 2]
```diff
@@ -68,17 +67,27 @@ cargo install diesel_cli --no-default-features --features "sqlite-bundled postgr
 - Migrate database tables:
 
 ```bash
+# The following commands assume you are in the server directory
+cd server
```
Comment (andreievg) [inline README.md:71]: why is this needed, i ran `diesel migration run --database-url ./omsupply-database.sqlite --migration-dir ../repository/migrations/sqlite/` from root repository and it created `omsupply-database.sqlite`

File: graphql/graphql_schema/export.rs [hunk 3]
```diff
@@ -0,0 +1,21 @@
+use graphql::build_schema;
+use std::fs;
+//  Entry point for export_graphql target, see cargo.toml and reademe.md test section
+fn main() {
+    // Try to detect if we are in the graphql crate or in the root crate and adjust the output
+    // path accordingly.
+    let has_schema_dir = fs::read_dir(".").unwrap().find_map(|entry| {
```
Comment (andreievg) [inline graphql/graphql_schema/export.rs:7]: hmm, can we assume that we always run from root ? And that's the only way to run the project crates ?

File: README.md (hunk 2 above)
Comment (clemens-msupply) [reply README.md:71]: You mean you called it from repository? yes that works as well, it just doesn't work if you are in the workspace root.

File: server/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,42 @@
+[package]
+name = "remote_server"
```
Comment (andreievg) [inline server/Cargo.toml:2]: i think package name should be the same as workspace package name ?

Also should configurations be in root, rather in this crate ?

**Review by andreievg — COMMENTED** [review]
Thanks ! That's huge effort. 

I want to test before approving. 

Main comment for me is, i think we should alway run everything from `root`, rather then from individual crates.

File: README.md (hunk 2 above)
Comment (andreievg) [reply README.md:71]: I ran it from workspace root

Comment (andreievg) [reply README.md:71]: oops, i didn't paste the correct command
`diesel migration run --database-url ./omsupply-database.sqlite --migration-dir ./repository/migrations/sqlite/`

File: repository/src/repository/diesel/item_query.rs (hunk 1 above)
Comment (andreievg) [reply repository/src/repository/diesel/item_query.rs:76]: > Or would comment help?  

I think if we can do it with a method name, there is no need for comment. (don't need to maintain the comment or validate it)

> Actually like it because the whole method is not to big

I am wondering, what would be considered big ? To me invoice query is big enough to warrant extraction of that method. And in that case it's better to use a consistent pattern across all domain repos

I really like the practice of extraction logic into method that explain what the logic is with the method name, and letting the dev have an option to dig deeper into the logic if they need to, or just see the overview, rather then forcing them to parse all of the code in their head to deduce the meaning.

**Comment by andreievg** [comment]
> EDIT: I also removed the "mock" feature, should I put it back in? Was late last night and I spontaneously decided we are not using it and it can be removed.

I think mock features is not needed

**Comment by andreievg** [comment]
> update graphql_export so that it can run from the project root and from the graphql crate (it tries to auto detect from where it is running and where to generate the output, thought this could avoids some confusion...)

As per earlier comment, why would we want to run things from individual crates ?

**Comment by andreievg** [comment]
> Removed RepositoryRegistry and only store the StorageConnectionManager in the app data (this simplified the workspace transition and RepositoryRegistry was only holding StorageConnectionManager anyway...)

great

**Comment by andreievg** [comment]
> Inline a handful of froms (3 or 4 if I remember correctly) They caused errors like below and it was easier to inline them than to workaround the tuple problem:

The more i see inline code, the less i question the notion that it's more readable. As per earlier comments if the method name describes correctly what the logic does, there should be no need to look into the actual underlying code. I think in general it's easier to digest code in 'layers', rather than large chunks with all of the layers.

File: graphql/src/lib.rs [hunk 5]
```diff
@@ -1,3 +1,12 @@
+// where_clauses_object_safety is a future compatibility lint over a soundness issue which is
```
Comment (andreievg) [inline graphql/src/lib.rs:1]: Was this comment added in this PR ? Wondering where does the annoying notification appear ? I tried removing `#![allow(where_clauses_object_safety)]` globally, and had no warnings in compilation

File: repository/src/repository/diesel/item_query.rs (hunk 1 above)
Comment (clemens-msupply) [reply repository/src/repository/diesel/item_query.rs]: done in 9910caa

Just bike-shedding: big depends a bit on how much logic is in the method. In this case it just a simple DB read for which reason I would say its small (just verbose) and have left it inline. If to_domain has some more transform logic in it and is used multiple times I would put it in a method.

File: graphql/src/lib.rs (hunk 5 above)
Comment (clemens-msupply) [reply graphql/src/lib.rs]: This was already in the old code. Maybe its not valid anymore, removed in daa9382

File: server/Cargo.toml (hunk 4 above)
Comment (clemens-msupply) [reply server/Cargo.toml]: done in b477daf

File: graphql/graphql_schema/export.rs (hunk 3 above)
Comment (clemens-msupply) [reply graphql/graphql_schema/export.rs:7]: Agree only problem I see is. How do we distribute the final binary and how do we expect the dir structure to be. As you said below, we can move configurations dir into the root and everything should work fine :) Will try it this afternoon.

**Review by andreievg — APPROVED** [review]
Thanks for changes looks good!

**Comment by andreievg** [comment]
Oops i didn't do a full check after approving. Pending `everything should be run from root` comment. 

If you run with `-p` it would look inside configuration folder in root of the project, which i think makes it work correctly in producation

**Comment by clemens-msupply** [comment]
All commands assume the root directory now: ef9727c

**Comment by andreievg** [comment]
> All commands assume the root directory now:

Thanks, just one more thing to pass tests, 

https://github.com/openmsupply/remote-server/compare/558-workspaces...558-tests
